kyuubi-common/src/main/scala/org/apache/kyuubi/operation/OperationManager.scala

```diff
@@ -189,4 +192,30 @@ abstract class OperationManager(name: String) extends AbstractService(name) {
       null
     }
   }
+
+  def waitForOperationsToFinish(): Unit = {
+    val timeWhenStopStarted = System.nanoTime()
+    val stopTimeoutMs = conf.get(SESSION_CLOSE_GRACEFULLY_TIMEOUT)
+    val pollTime = 300
+
+    // To prevent graceful stop to get stuck permanently
+    def hasTimedOut: Boolean = {
+      val diff = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - timeWhenStopStarted)
+      val timedOut = diff > stopTimeoutMs
+      if (timedOut) {
+        warn("Timed out while waiting the spark operations to finish " +
+          "(timeout = " + stopTimeoutMs + ")")
+      }
+      timedOut
+    }
+
+    info("Waiting for all the operations to finish.")
+    while (!hasTimedOut && !allOperations().filter(op =>
```

Comment:
`allOperations().exists(...)` to avoid transform tail operations if has active operation.